lib/elixir/lib/module/checker.ex

```diff
@@ -0,0 +1,130 @@
+defmodule Module.Checker do
+  def verify(module_map, _binary) do
+    state = :maps.with([:module, :file, :compile_opts], module_map)
+
+    module_map.definitions
+    |> Enum.reverse()
+    |> check_definitions(state)
+    |> List.flatten()
+  end
+
+  defp check_definitions(definitions, state) do
+    Enum.map(definitions, &check_definition(&1, state))
+  end
+
+  defp check_definition({def, _kind, meta, clauses}, state) do
+    state = :maps.put(:function, def, put_file_meta(state, meta))
+    Enum.map(clauses, &check_clause(&1, state))
+  end
+
+  defp put_file_meta(state, meta) do
+    case Keyword.fetch(meta, :file) do
+      {:ok, {file, line}} ->
+        :maps.merge(state, %{file: file, line: line})
+
+      :error ->
+        state
+    end
+  end
+
+  defp check_clause({_meta, _args, _guards, body}, state) do
+    check_body(body, state)
+  end
+
+  defp check_body({:&, _, [{:/, _, [{{:., meta, [module, fun]}, _, []}, arity]}]}, state)
+       when is_atom(module) and is_atom(fun) do
+    check_remote(module, fun, arity, meta, state)
+  end
+
+  defp check_body({{:., meta, [module, fun]}, _, args}, state)
+       when is_atom(module) and is_atom(fun) do
+    check_remote(module, fun, length(args), meta, state)
+  end
+
+  defp check_body({left, _meta, right}, state) when is_list(right) do
+    [check_body(right, state), check_body(left, state)]
+  end
+
+  defp check_body({left, right}, state) do
+    [check_body(right, state), check_body(left, state)]
+  end
+
+  defp check_body(list, state) when is_list(list) do
+    Enum.map(list, &check_body(&1, state))
+  end
+
+  defp check_body(_other, _state) do
+    []
+  end
+
+  defp check_remote(module, fun, arity, meta, state) do
+    cond do
+      not should_warn_undefined?(module, fun, arity, state) ->
+        []
+
+      # TODO: In the future we may want to warn for modules defined
+      # in the local context
+      Keyword.get(meta, :context_module, false) and state.module != module ->
+        []
+
+      # TODO: Add no_autload
+      not Code.ensure_loaded?(module) ->
+        warn(meta, state, {:undefined_module, module, fun, arity})
+
+      not function_exported?(module, fun, arity) ->
+        exports = exports_for(module)
+        warn(meta, state, {:undefined_function, module, fun, arity, exports})
+
+      true ->
+        []
+    end
+  end
+
+  # TODO: Do not warn inside guards
+  # TODO: Properly handle protocols
+  defp should_warn_undefined?(_module, :__impl__, 1, _state), do: false
```

Comment:
Nitpick: I think the `should_` could be dropped:
```suggestion
  defp warn_undefined?(_module, :__impl__, 1, _state), do: false
```

Reply:
It may not warn so I think it is also confusing. The intention behind the name is that the function should warn if the other checks also fail. Do you know a better name?

Reply:
`warn_if_undefined?` maybe?